internal/runtime/core.go

```diff
@@ -44,18 +47,63 @@ func NewServiceAccount(namespace, name string) *corev1.ServiceAccount {
 }
 
 // NewSecret returns a corev1.Secret with namespace and name.
-func NewSecret(namespace, name string, data map[string][]byte) *corev1.Secret {
+func NewSecret(namespace, name string, data map[string][]byte, visitors ...func(meta metav1.Object)) *corev1.Secret {
 	if data == nil {
 		data = map[string][]byte{}
 	}
 	s := &corev1.Secret{Data: data}
-	Initialize(s, namespace, name)
+	Initialize(s, namespace, name, visitors...)
 	return s
 }
 
-// NewDaemonSet returns a daemon set.
-func NewDaemonSet(namespace, name string) *appsv1.DaemonSet {
-	ds := &appsv1.DaemonSet{}
-	Initialize(ds, namespace, name)
+func NewDaemonSet(daemonSetName, namespace, component string, podSpec corev1.PodSpec, visitors ...func(meta metav1.Object)) *appsv1.DaemonSet {
```

Comment:
@alanconway how much do you care about this refactoring with the change in args? My initial understanding was that:

* runtime - Place for "generic" runtime factories
* collector/.... - place for collector specific factories

This change is combining the two to do away with one package that was taken over by 'runtime' .  @vparfonov seems like:

* the label changes should be pushed into the label initializer you introduced
* the DS specific template, scheduling, rollout strategy could be pushed into collector specific DS

Reply:
I like the visitors... approach. It is a very common way to express "options" in Go. It doesn't change the generic nature of runtime, but it lets the caller inject additional setup on the fly. Not sure I understood your question @jcantrill